src/os/bluestore/BlueStore.cc

```diff
@@ -13623,6 +13692,13 @@ int BlueStore::_do_alloc_write(
   }
   _collect_allocation_stats(need, min_alloc_size, prealloc.size());
 
+  if (bdev->is_smr()) {
+    std::lock_guard l{zoned_cleaner_lock};
```

Comment:
I'm a bit uncomfortable with the need to unconditionally take this lock...
May be use some intermediate container to get zones to clean to. And then move/append its content under the appropriate lock to zoned_cleaner_queue if needed.
Less straightforward way would be to pass zoned_cleaner_lock inside zone_get_zones_to_clean and take it there if needed...

Reply:
I've addressed it using the first method you suggested, please take a look.  I'm not very happy about this because now we get a lock inside zoned_get_zones_to_clean, and then we get it again here to populate zoned_cleaner_queue.  So perhaps having two different functions is a better way?  One to check if there is a need to clean, which only reads num_free, and if there is, we acquire lock once and populate zoned_cleaner_queue.  What do you think?

Reply:
You'll need two locks acquisition (for two different locks!) anyway. One to be acquired for allocator and another one for cleanup thread. The difference is only whether we need to take both locks simultaneously or one-by-one. The latter IMO is preferred. And that's what you have for now.